design/parallel-commit/parallel-commit-solution-ideas.md

```diff
@@ -0,0 +1,59 @@
+
+We cannot asynchronously allocate a TSO as transaction `T`'s commit_ts after telling the client that the transaction has been finished, because it's possible that the client runs faster and got a earlier TSO to start its next transaction `T2`, so that in `T2`'s sight the previous transaction `T` didn't commit.
+
+Neither can we tell the client `T` has been committed just after allocating TSO as `T`'s commit_ts. Because if server crashes it will never know what TS it has allocated, and it can not find the proper commit_ts anymore.
+
+Actually, we believe we should persist some information that helps us in finding the commit_ts, and it should have been persisted when the transaction has been "successfully committed". Our current idea is persist `max_read_ts` into the lock that we prewrites, and the final commit_ts will be `max_over_all_keys{max_read_ts}+1`. However it's hard: we need to get the current value of `max_read_ts` before writing down the locks, however new reads may happens between getting `max_read_ts` and successfully writing down the lock.
+
+We may have no perfect solution to this problem, but we have came up with some different ideas that may be possible to solve it.
+
+### Idea 1: Memory Locking 
+
+When we got the max read ts as `M` and then prewriting key `K` with max read ts `M`, we want to prevent other transactions from reading `K` with a ts larger than `M`. The most straightforward way is to make a in-memory lock: A Parallel-Commit prewrite should lock the keys it want to write, before reading the `max_read_ts`, and release the lock after finishing prewrite. A read operation should check the memory lock after updating `max_read_ts` and should be blocked (or return a KeyIsLocked error) when it meets the memory lock. Read operations don't need to actually acquire locks. It can proceed reading if it finds that the keys it wants to read are not locked.
+
+The memory lock should support range-based queries, because a reading operation may need to scan a range. So the memory lock should be implemented by an ordered map (like BTreeMap, Trie, SkipList).
+
+Therefore, the major difficulty of memory lock solution is that how to keep the performance.
+
+### Idea 2: RocksDB-Based Lock (By @gengliqi)
+
+This idea differs from [Idea1](#idea-1-memory-locking), only in that we are using RocksDB instead of an in-memory data structure to store locks. We avoid most performance impact if we can put it directly to LOCK_CF without Raft replication, however, it might introduce too many corner cases and troubles to resolve, considering that there are leader changing and region scheduling in TiKV. Therefore we may add a new CF to store these kind of locks. The performance comparing to [Idea1](#idea-1-memory-locking) is uncertain before we actually do a POC test.
+
+### Idea 3: Transaction List (By @Little-Wallace)

```

Comment:
Sorry but I don't fully understand what's the difference between idea 1 and 3. I think essentially speaking they are all locks in memory.

Reply:
In Idea 1 the locks are managed by a map of keys, and reads  need to check keys in it. In idea 3 we keeps a list of running prewrites and reads need to check each items in the list who  have got a smaller `max_read_ts` than the current read transaction.

Reply:
So if I understand it correctly, comparing Idea 1 and 3 is actually weighing the cost of searching keys in the map and iterating over the running transactions?

Reply:
Yes. @Little-Wallace is trying to make a optimized design of the transaction list.